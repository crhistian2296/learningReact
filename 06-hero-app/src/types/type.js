const type = {
  login: '[auth]: Login',
  logout: '[auth]: Logout',
};

export const { login, logout } = type;
