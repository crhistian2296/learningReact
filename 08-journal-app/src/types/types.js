// export const type = {
//   login: '[Auth] Login',
//   logout: '[Auth] Logout',
// };

export const login = '[Auth] Login',
  logout = '[Auth] Logout';
