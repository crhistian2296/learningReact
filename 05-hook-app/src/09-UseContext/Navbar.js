import { NavLink } from 'react-router-dom';

const Navbar = () => {
  return (
    //* Navbar cogida de getBootstrap
    <nav className='navbar navbar-expand navbar-dark bg-dark'>
      <div className='container-fluid'>
        <NavLink className='navbar-brand' to='/'>
          useContext
        </NavLink>
        <div className='collapse navbar-collapse' id='navbarNav'>
          <ul className='navbar-nav'>
            <NavLink className='nav-item nav-link ' to='/'>
              Home
            </NavLink>
            <NavLink className='nav-item nav-link ' to='/about'>
              About
            </NavLink>
            <NavLink className='nav-item nav-link ' to='/login'>
              Login
            </NavLink>
          </ul>
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
