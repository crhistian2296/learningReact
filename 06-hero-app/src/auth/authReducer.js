// const user = {
//   name:'Fernando',
//   logged: true
// }

import { login, logout } from '../types/type';

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case login:
      return {
        ...action.payload,
        logged: true,
      };
    case logout:
      return {
        logged: false,
      };
    default:
      return state;
  }
};
