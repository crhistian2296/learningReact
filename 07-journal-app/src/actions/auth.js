import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { uiFinishLoading, uiStartLoading } from './ui';

// export const startLoginEmailPassword = (email, password) => {
//   // *redux-thunk provee automaticamente ese dispatch para acciones asincronas
//   return (dispatch) => {
//     setTimeout(() => {
//       dispatch(login(123, 'crhistian'));
//     }, 3000);
//   };
// };

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
/**
 * Crea una cuenta en la app de autenticacion en firebase mediante correo
 * @param {string} email
 * @param {string} password
 * @param {string} name
 * @returns void
 */
export const startRegisterWithEmailAsync = (email, password, name) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};
export const startLoginWithEmailAsync = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};

/**
 * Crean una cuenta en la app de autenticacion de firebase mediante boton de google
 * @returns void
 */
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch(console.log);
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    await dispatch(logout());
  };
};

export const logout = () => ({ type: types.logout });
