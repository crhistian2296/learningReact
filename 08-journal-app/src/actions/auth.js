import { login } from '../types/types';

export const loginAction = (uid, displayName) => ({
  type: login,
  payload: {
    uid,
    displayName
  }
})
