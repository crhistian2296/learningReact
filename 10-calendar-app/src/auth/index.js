export * from './pages/Login.jsx';
