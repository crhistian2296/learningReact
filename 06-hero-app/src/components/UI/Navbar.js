// Navbar tomado de la documentacion de Bootstrap
import { useContext } from 'react';
import { Link, NavLink, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { logout } from '../../types/type';

export const Navbar = () => {
  const navigate = useNavigate();
  const { dispatch } = useContext(AuthContext);

  const handleLogout = () => {
    dispatch({
      type: logout,
    });
    navigate('/login', { replace: true });
  };

  const { user } = useContext(AuthContext);

  return (
    <nav className='navbar navbar-expand-sm navbar-dark bg-dark'>
      <div className='container-fluid'>
        <Link className='navbar-brand' to='/'>
          Asociaciones
        </Link>

        <div className='navbar-collapse justify-content-sm-between '>
          <div className='navbar-nav'>
            <NavLink className='nav-item nav-link' to='/marvel'>
              Marvel
            </NavLink>
            <NavLink className='nav-item nav-link' to='/dc'>
              DC
            </NavLink>
            <NavLink className='nav-item nav-link' to='/search'>
              Search
            </NavLink>
          </div>
          <div className='navbar-nav'>
            <span className='nav-item nav-link text-info'>{user.name}</span>
            <button className='nav-item nav-link btn logoutLink' onClick={handleLogout}>
              Logout
            </button>
          </div>
        </div>
      </div>
    </nav>
  );
};
