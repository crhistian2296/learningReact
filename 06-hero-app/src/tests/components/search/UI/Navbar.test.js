import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AuthContext } from '../../../../auth/AuthContext';
import { Navbar } from '../../../../components/UI/Navbar';
import { logout } from '../../../../types/type';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas sobre <Navbar/>', () => {
  const contextValue = {
    user: {
      logged: true,
      name: 'Crhistian',
    },
    dispatch: jest.fn(),
  };
  let wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path='/' element={<Navbar />}></Route>
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path='/' element={<Navbar />}></Route>
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    );
  });

  test('Debe renderizarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.text-info').text().trim()).toBe('Crhistian');
  });

  test('Debe de llamar el logout, el navigate y el dispatch con los argumentos', () => {
    const { dispatch } = contextValue;

    wrapper.find('.logoutLink').simulate('click');

    expect(dispatch).toHaveBeenCalledWith({ type: logout });
    expect(mockNavigate).toHaveBeenCalledWith('/login', { replace: true });
  });
});
