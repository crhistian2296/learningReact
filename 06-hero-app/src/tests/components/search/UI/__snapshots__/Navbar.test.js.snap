// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <Navbar/> Debe renderizarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse justify-content-sm-between "
    >
      <div
        className="navbar-nav"
      >
        <a
          className="nav-item nav-link"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          className="nav-item nav-link"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-item nav-link"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
      <div
        className="navbar-nav"
      >
        <span
          className="nav-item nav-link text-info"
        >
          Crhistian
        </span>
        <button
          className="nav-item nav-link btn logoutLink"
          onClick={[Function]}
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</nav>
`;
